### PR TITLE
ansible-galaxy - fix collection installation with trailing slashes

### LINKED ISSUE #70009: `ansible-galaxy collection install` explodes for some collections on `devel`
#### SUMMARY

Installing `amazon.aws` collection fails on ansible/ansible@devel, supposedly because of the embedded symlinks:
```console
ERROR! Unexpected Exception, this is probably a bug: "filename 'tests/unit/mod
ules/placebo_recordings/aws_direct_connect_link_aggregation_group/nonexistent_
lag_status' not found"
```

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

ansible-galaxy CLI

##### ANSIBLE VERSION

devel

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

N/A

##### STEPS TO REPRODUCE

Running the install should be enough:

```console
$ ANSIBLE_DEBUG=1 ansible-galaxy -vvvv collection install amazon.aws
```

##### EXPECTED RESULTS

The collection install should succeed.

##### ACTUAL RESULTS

```python-traceback
Exception, this is probably a bug: "filename 'tests/unit/modules/placebo_recordings/aws_direct_connect_link_aggregation_group/nonexistent_lag_status' not found"
the full traceback was:
Traceback (most recent call last):
  File "/home/bcoca/work/ansible/bin/ansible-galaxy", line 123, in <module>
    exit_code = cli.run()
  File "/home/bcoca/work/ansible/lib/ansible/cli/galaxy.py", line 490, in run
    context.CLIARGS['func']()
  File "/home/bcoca/work/ansible/lib/ansible/cli/galaxy.py", line 1057, in execute_install
    self._execute_install_collection(collection_requirements, collection_path)
  File "/home/bcoca/work/ansible/lib/ansible/cli/galaxy.py", line 1079, in _execute_install_collection
    no_deps, force, force_with_deps, allow_pre_release=allow_pre_release)
  File "/home/bcoca/work/ansible/lib/ansible/galaxy/collection.py", line 675, in install_collections
    collection.install(output_path, b_temp_path)
  File "/home/bcoca/work/ansible/lib/ansible/galaxy/collection.py", line 250, in install
    self.install_artifact(b_collection_path, b_temp_path)
  File "/home/bcoca/work/ansible/lib/ansible/galaxy/collection.py", line 275, in install_artifact
    _extract_tar_dir(collection_tar, file_name, b_collection_path)
  File "/home/bcoca/work/ansible/lib/ansible/galaxy/collection.py", line 1387, in _extract_tar_dir
    tar_member = tar.getmember(to_native(dirname, errors='surrogate_or_strict'))
  File "/usr/lib/python3.7/tarfile.py", line 1752, in getmember
    raise KeyError("filename %r not found" % name)
KeyError: "filename 'tests/unit/modules/placebo_recordings/aws_direct_connect_link_aggregation_group/nonexistent_lag_status' not found"
```

### PR DESCRIPTION
**Pull request by samdoran**
##### SUMMARY

If we fail to find a member when extracting a directory, try adding a trailing slash to the member name. In certain cases, the member in the tarfile will contain a trailing slash but the file name in FILES.json will never contain the trailing slash.

If unable to find the member, handle the KeyError and print a nicer error.

Also check if a directory exists before creating it since it may have been extracted from the archive.

Fixes #70009

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

`lib/ansible/galaxy/collection.py`
##### ADDITIONAL INFORMATION

Not sure we need a changelog since this was just added yesterday.

I'm open to suggestions for tests on this. It would be much easier to write unit tests rather than integration tests since I'm not sure how to create tar file where some members contain a trailing slash. If unit tests are ok, I'll work on adding those.

### CODE REVIEW

File: lib/ansible/galaxy/collection.py
Comment (webknjaz) [inline lib/ansible/galaxy/collection.py]: Currently, you just retry the same thing again if there's a `/` already. Maybe you should drop it?
```suggestion
        # Some archives contain directories that end with '/'
        if not dirname.endswith(os.path.sep):
            dirname += os.path.sep
        else:
            dirname = dirname[:-1]
```

File: lib/ansible/galaxy/collection.py
Comment (webknjaz) [inline lib/ansible/galaxy/collection.py]: This should probably mention both slash and no-slash paths...

File: lib/ansible/galaxy/collection.py
Comment (jborean93) [inline lib/ansible/galaxy/collection.py]: I don't want to hold up the PR but you could use a loop instead of a nested KeyError like

```suggestion
    member_names = [to_native(dirname, errors='surrogate_or_strict')]
    if not member_names[-1].endswith(os.path.sep):
        member_names.append(member_names[-1] + os.path.sep)
        
    for name in member_names:
        try:
            tar_member = tar.getmember(name)
        except KeyError:
            continue
        break
    else:
        raise AnsibleError("Unable to extract '%s' from collection" % member_names[0])
```

Potentially allows us to try more options in the future if need be.

**Review by jborean93 — COMMENTED** [review]
Thanks for working on this, no idea how the aws collection was set up like this but will definitely need to fix it. Once this is merged we will have to backport it to stable-2.9 as well.

File: lib/ansible/galaxy/collection.py
Comment (samdoran) [reply lib/ansible/galaxy/collection.py]: Good point. Most of the items do not have a trailing slash. There's room for optimization here. I may end up doing a loop as @jborean93 suggested below.

File: lib/ansible/galaxy/collection.py
Comment (samdoran) [reply lib/ansible/galaxy/collection.py]: Since it seems this archive with the trailing `/` is an anomaly, I don't think we need to mention both versions of the path.

File: lib/ansible/galaxy/collection.py
Comment (samdoran) [reply lib/ansible/galaxy/collection.py]: I need to head out for today. I will try to poke at this a bit tomorrow even though it's technically a holiday. If not, I will get to it on Monday.

Comment (samdoran) [reply lib/ansible/galaxy/collection.py]: I ended up going with this solution. I'm trying to work out why it's not breaking out of the loop and adjusting the tests.

Comment (samdoran) [reply lib/ansible/galaxy/collection.py]: Found my bug. 😊

**Review by jborean93 — APPROVED** [review]

**Review by s-hertel — APPROVED** [review]

**Comment by samdoran** [comment]
I'll work on backporting this.

**Comment by samdoran** [comment]
This actually doesn't seem to be broken in `stable-2.9` since the `_extract_tar_dir()` method does not exist in that branch, nor does the commit, d30fc6c0b359f631130b0e979d9a78a7b3747d48, which introduced this bug. The problematic `tar` file extracts with with `stable-2.9`. Do we need to backport this then, @s-hertel @jborean93?

**Comment by samdoran** [comment]
I attempted to backport this change and was going through the merge conflict resolution thinking _a lot_ has changed in that file. If we want to backport this fix, I think we would also have to backport new code using `_extract_tar_dir()`.

**Comment by jborean93** [comment]
@samdoran looks like the backport PR hasn't been merged yet sorry https://github.com/ansible/ansible/pull/69994. I thought @nitzmahone had already done that. Are you able to add your backport onto that?

**Comment by jborean93** [comment]
I've backported the change to https://github.com/ansible/ansible/pull/69994 so they all get merged into stable-2.9 at once.

**Comment by samdoran** [comment]
@jborean93 That explains it. I didn't know we were planning to backport #69994, which is why I was confused. Thanks for adding it to that PR.

**Comment by jborean93** [comment]
@samdoran it's unfortunately needed to ensure 2.9 will work with symlink dirs and actually preserve the symlink files. Technically it isn't needed as 2.9 will continue to install collections that contain symlinks but not in a nice way. The biggest driver for adding symlinks was ensuring the sanity tests pass for symlinked modules which was used for redirection before redirection was a thing on 2.9 hence why they needed this there.

**Pull request merged**
